File: modyn/storage/internal/database/storage_database_connection.py
"""Database connection context manager."""

from __future__ import annotations

import logging

from modyn.database.abstract_database_connection import AbstractDatabaseConnection
from modyn.storage.internal.database.models import Dataset, File, Sample
from modyn.storage.internal.database.storage_base import StorageBase
from modyn.storage.internal.file_wrapper.file_wrapper_type import FileWrapperType
from modyn.storage.internal.filesystem_wrapper.filesystem_wrapper_type import FilesystemWrapperType
from sqlalchemy import exc

logger = logging.getLogger(__name__)


class StorageDatabaseConnection(AbstractDatabaseConnection):
    """Database connection context manager."""

    def __init__(self, modyn_config: dict) -> None:
        """Initialize the database connection.

        Args:
            modyn_config (dict): Configuration of the modyn module.
        """
        super().__init__(modyn_config)
        self.drivername: str = self.modyn_config["storage"]["database"]["drivername"]
        self.username: str = self.modyn_config["storage"]["database"]["username"]
        self.password: str = self.modyn_config["storage"]["database"]["password"]
        self.host: str = self.modyn_config["storage"]["database"]["host"]
        self.port: int = self.modyn_config["storage"]["database"]["port"]
        self.database: str = self.modyn_config["storage"]["database"]["database"]
        self.hash_partition_modulus: int = (
            self.modyn_config["storage"]["database"]["hash_partition_modulus"]
            if "hash_partition_modulus" in self.modyn_config["storage"]["database"]
            else 8
        )
        self.sample_table_unlogged: bool = (
            self.modyn_config["storage"]["database"]["sample_table_unlogged"]
            if "sample_table_unlogged" in self.modyn_config["storage"]["database"]
            else True
        )

    def __enter__(self) -> StorageDatabaseConnection:
        """Create the engine and session.

        Returns:
            DatabaseConnection: DatabaseConnection.
        """
        super().__enter__()
        return self

    def create_tables(self) -> None:
        """
        Create all tables. Each table is represented by a class.

        All classes that inherit from Base are mapped to tables
        which are created in the database if they do not exist.

        The metadata is a collection of Table objects that inherit from Base and their associated
        schema constructs (such as Column objects, ForeignKey objects, and so on).
        """
        Sample.ensure_pks_correct(self.session)
        StorageBase.metadata.create_all(self.engine)

    def add_dataset(
        self,
        name: str,
        base_path: str,
        filesystem_wrapper_type: FilesystemWrapperType,
        file_wrapper_type: FileWrapperType,
        description: str,
        version: str,
        file_wrapper_config: str,
        ignore_last_timestamp: bool = False,
        file_watcher_interval: int = 5,
    ) -> bool:
        """
        Add dataset to database.

        If dataset with name already exists, it is updated.
        """
        try:
            if self.session.query(Dataset).filter(Dataset.name == name).first() is not None:
                logger.info(f"Dataset with name {name} exists.")
                self.session.query(Dataset).filter(Dataset.name == name).update(
                    {
                        "base_path": base_path,
                        "filesystem_wrapper_type": filesystem_wrapper_type,
                        "file_wrapper_type": file_wrapper_type,
                        "description": description,
                        "version": version,
                        "file_wrapper_config": file_wrapper_config,
                        "ignore_last_timestamp": ignore_last_timestamp,
                        "file_watcher_interval": file_watcher_interval,
                    }
                )
            else:
                logger.info(f"Dataset with name {name} does not exist.")
                dataset = Dataset(
                    name=name,
                    base_path=base_path,
                    filesystem_wrapper_type=filesystem_wrapper_type,
                    file_wrapper_type=file_wrapper_type,
                    description=description,
                    version=version,
                    file_wrapper_config=file_wrapper_config,
                    last_timestamp=-1,  # Set to -1 as this is a new dataset
                    ignore_last_timestamp=ignore_last_timestamp,
                    file_watcher_interval=file_watcher_interval,
                )
                self.session.add(dataset)
            self.session.commit()
        except exc.SQLAlchemyError as exception:
            logger.error(f"Error adding dataset: {exception}")
            self.session.rollback()
            return False
        return True

    def delete_dataset(self, name: str) -> bool:
        """Delete dataset from database."""
        try:
            self.session.query(Sample).filter(
                Sample.file_id.in_(self.session.query(File.file_id).join(Dataset).filter(Dataset.name == name))
            ).delete(synchronize_session="fetch")
            self.session.query(File).filter(
                File.dataset_id.in_(self.session.query(Dataset.dataset_id).filter(Dataset.name == name))
            ).delete(synchronize_session="fetch")
            self.session.query(Dataset).filter(Dataset.name == name).delete(synchronize_session="fetch")
            self.session.commit()
        except exc.SQLAlchemyError as exception:
            logger.error(f"Error deleting dataset: {exception}")
            self.session.rollback()
            return False
        return True

    def add_sample_dataset(self, dataset_id: int) -> None:
        """Add a new dataset to the samples table.

        This method creates a new partitions for the dataset.

        Args:
            dataset_id (int): Id of the dataset
        """
        Sample.add_dataset(
            dataset_id, self.session, self.engine, self.hash_partition_modulus, self.sample_table_unlogged
        )

File: modyn/storage/src/internal/database/storage_database_connection.cpp
#include "internal/database/storage_database_connection.hpp"

#include <fmt/format.h>
#include <spdlog/spdlog.h>

#include <filesystem>
#include <fstream>

#include "modyn/utils/utils.hpp"
#include "soci/postgresql/soci-postgresql.h"
#include "soci/sqlite3/soci-sqlite3.h"

using namespace modyn::storage;

soci::session StorageDatabaseConnection::get_session() const {
  const std::string connection_string =
      fmt::format("dbname={} user={} password={} host={} port={}", database_, username_, password_, host_, port_);
  soci::connection_parameters parameters;

  switch (drivername_) {
    case DatabaseDriver::POSTGRESQL:
      parameters = soci::connection_parameters(soci::postgresql, connection_string);
      break;
    case DatabaseDriver::SQLITE3:
      parameters = soci::connection_parameters(soci::sqlite3, connection_string);
      break;
    default:
      FAIL("Unsupported database driver");
  }
  return soci::session(parameters);
}

void StorageDatabaseConnection::create_tables() const {
  soci::session session = get_session();

  std::string dataset_table_sql;
  std::string file_table_sql;
  std::string sample_table_sql;
  switch (drivername_) {
    case DatabaseDriver::POSTGRESQL:
      dataset_table_sql =
#include "sql/PostgreSQLDataset.sql"
          ;
      file_table_sql =
#include "sql/PostgreSQLFile.sql"
          ;
      sample_table_sql =
#include "sql/PostgreSQLSample.sql"
          ;
      break;
    case DatabaseDriver::SQLITE3:
      dataset_table_sql =
#include "sql/SQLiteDataset.sql"
          ;
      file_table_sql =
#include "sql/SQLiteFile.sql"
          ;
      sample_table_sql =
#include "sql/SQLiteSample.sql"
          ;
      break;
    default:
      FAIL("Unsupported database driver");
  }
  session << dataset_table_sql;

  session << file_table_sql;

  session << sample_table_sql;

  if (drivername_ == DatabaseDriver::POSTGRESQL && sample_table_unlogged_) {
    session << "ALTER TABLE samples SET UNLOGGED";
  }

  session.close();
}

bool StorageDatabaseConnection::add_dataset(const std::string& name, const std::string& base_path,
                                            const FilesystemWrapperType& filesystem_wrapper_type,
                                            const FileWrapperType& file_wrapper_type, const std::string& description,
                                            const std::string& version, const std::string& file_wrapper_config,
                                            const bool ignore_last_timestamp, const int64_t file_watcher_interval,
                                            const bool upsert) const {
  soci::session session = get_session();

  auto filesystem_wrapper_type_int = static_cast<int64_t>(filesystem_wrapper_type);
  auto file_wrapper_type_int = static_cast<int64_t>(file_wrapper_type);
  std::string boolean_string = ignore_last_timestamp ? "true" : "false";

  const int64_t dataset_id = get_dataset_id(name);

  if (dataset_id != -1 && !upsert) {
    // Dataset already exists
    SPDLOG_ERROR("Cannot insert dataset {} as it already exists", name);
    return false;
  }

  if (dataset_id != -1) {
    SPDLOG_INFO("Dataset {} already exists, updating...", name);

    const std::string update_query =
#include "sql/SQLUpdateDataset.sql"
        ;

    session << update_query, soci::use(base_path), soci::use(filesystem_wrapper_type_int),
        soci::use(file_wrapper_type_int), soci::use(description), soci::use(version), soci::use(file_wrapper_config),
        soci::use(boolean_string), soci::use(file_watcher_interval), soci::use(dataset_id);
    return true;
  }

  // Dataset does not exist
  const std::string insertion_query =
#include "sql/SQLInsertDataset.sql"
      ;

  switch (drivername_) {
    // same logic for both dbms > fallthrough
    case DatabaseDriver::POSTGRESQL:
    case DatabaseDriver::SQLITE3:
      session << insertion_query, soci::use(name), soci::use(base_path), soci::use(filesystem_wrapper_type_int),
          soci::use(file_wrapper_type_int), soci::use(description), soci::use(version), soci::use(file_wrapper_config),
          soci::use(boolean_string), soci::use(file_watcher_interval);
      break;
    default:
      SPDLOG_ERROR("Error adding dataset: Unsupported database driver.");
      return false;
  }

  // Create partition table for samples
  if (!add_sample_dataset_partition(name)) {
    FAIL("Partition creation failed.");
  }
  return true;
}

int64_t StorageDatabaseConnection::get_dataset_id(const std::string& name) const {
  soci::session session = get_session();

  int64_t dataset_id = -1;
  session << "SELECT dataset_id FROM datasets WHERE name = :name", soci::into(dataset_id), soci::use(name);
  session.close();

  return dataset_id;
}

DatabaseDriver StorageDatabaseConnection::get_drivername(const YAML::Node& config) {
  ASSERT(config["storage"]["database"], "No database configuration found");

  const auto drivername = config["storage"]["database"]["drivername"].as<std::string>();
  if (drivername == "postgresql") {
    return DatabaseDriver::POSTGRESQL;
  }
  if (drivername == "sqlite3") {
    return DatabaseDriver::SQLITE3;
  }

  FAIL("Unsupported database driver: " + drivername);
}

bool StorageDatabaseConnection::delete_dataset(const std::string& name, const int64_t dataset_id) const {
  soci::session session = get_session();

  // Delete all samples for this dataset
  try {
    session << "DELETE FROM samples WHERE dataset_id = :dataset_id", soci::use(dataset_id);
  } catch (const soci::soci_error& e) {
    SPDLOG_ERROR("Error deleting samples for dataset {}: {}", name, e.what());
    return false;
  }

  // Delete all files for this dataset
  try {
    session << "DELETE FROM files WHERE dataset_id = :dataset_id", soci::use(dataset_id);
  } catch (const soci::soci_error& e) {
    SPDLOG_ERROR("Error deleting files for dataset {}: {}", name, e.what());
    return false;
  }

  // Delete the dataset
  try {
    session << "DELETE FROM datasets WHERE name = :name", soci::use(name);
  } catch (const soci::soci_error& e) {
    SPDLOG_ERROR("Error deleting dataset {}: {}", name, e.what());
    return false;
  }

  session.close();

  return true;
}

bool StorageDatabaseConnection::add_sample_dataset_partition(const std::string& dataset_name) const {
  soci::session session = get_session();
  int64_t dataset_id = get_dataset_id(dataset_name);
  if (dataset_id == -1) {
    SPDLOG_ERROR("Dataset {} not found", dataset_name);
    return false;
  }
  switch (drivername_) {
    case DatabaseDriver::POSTGRESQL: {
      std::string dataset_partition_table_name = "samples__did" + std::to_string(dataset_id);
      try {
        session << fmt::format(
            "CREATE TABLE IF NOT EXISTS {} "
            "PARTITION OF samples "
            "FOR VALUES IN ({}) "
            "PARTITION BY HASH (sample_id)",
            dataset_partition_table_name, dataset_id);
      } catch (const soci::soci_error& e) {
        SPDLOG_ERROR("Error creating partition table for dataset {}: {}", dataset_name, e.what());
        session.close();
        return false;
      }

      try {
        for (int64_t i = 0; i < hash_partition_modulus_; i++) {
          std::string hash_partition_name = dataset_partition_table_name + "_part" + std::to_string(i);
          session << fmt::format(
              "CREATE TABLE IF NOT EXISTS {} "
              "PARTITION OF {} "
              "FOR VALUES WITH (modulus {}, REMAINDER {})",
              hash_partition_name, dataset_partition_table_name, hash_partition_modulus_, i);
        }
      } catch (const soci::soci_error& e) {
        SPDLOG_ERROR("Error creating hash partitions for dataset {}: {}", dataset_name, e.what());
        session.close();
        return false;
      }
      break;
    }
    case DatabaseDriver::SQLITE3: {
      SPDLOG_INFO(
          "Skipping partition creation for dataset {}, not supported for "
          "driver.",
          dataset_name);
      break;
    }
    default:
      FAIL("Unsupported database driver.");
  }

  session.close();
  return true;
}
